blogpost: Introducing a new database for code intel (#1784)
sourcegraph/about: blogposts/2020/introducing-codeintel-db.md

---
title: 'Introducing a new database for Code Intelligence'
author: 'Eric Fritz'
publishDate: 2020-10-18T00:00-07:00
tags: [
  "blog"
]
slug: introducing-a-new-code-intelligence-database
heroImage: https://image.flaticon.com/icons/png/512/51/51333.png
published: true
---

With the release of 3.21, Sourcegraph will ship with a _second_ Postgres database instance, named `codeintel-db`, which we use to store the data that powers [precise code intelligence](https://docs.sourcegraph.com/code_intelligence/explanations/precise_code_intelligence).

> We've [written previously](https://about.sourcegraph.com/blog/evolution-of-the-precise-code-intel-backend/) about the architectural changes we've made to the code intelligence backend. This post is yet another submission in the archaeological trail of how the code intelligence team's effort shaped Sourcegraph on a technical level, and will explain why we chose to swap out our persistence layer - a very non-trivial difference in our architecture.

Reviewer: ```suggestion
We've [written previously](https://about.sourcegraph.com/blog/evolution-of-the-precise-code-intel-backend/) about the architectural changes we have made to the code intelligence backend. This post is yet another submission in the archaeological trail of how the code intelligence team shapes Sourcegraph on a technical level. The post explains why we chose to swap out our persistence layer - a very non-trivial difference in our architecture.
```

## The impetus for a second database

A bit over a year ago, the code intelligence backend started as a single node process with access only to its own persistent volume. We chose SQLite as our persistence mechanism for a handful of reasons:

- It was the faster implementation compared to other contending designs we tried.
- Our MVP did not have access to Postgres at the time, which was owned solely by another team. Only later did we get access to Postgres to store cross-repository metadata and work queue data.
- SQLite had some nice properties such as easy data warehousing opportunities (on which we have not capitalized) and bulk deletion of expired data was straightforward. All we had to do was delete the SQLite file.

Using SQLite for this data has become an operational and feature burden for the code intelligence team.

1. We need to maintain a periodic process to reconcile the data on disk with the metadata we store in Postgres.
2. We have good instructions and tips for backup and recovery of our Postgres data store, but have never officially recommended backing up other persistent volumes.

> 3. We've locked ourselves into a node topology which is not embarrassingly scalable. We essentially have a singleton service over a persistent volume. In order to scale this service horizontally (which would become necessary once either the amount of data or the frequency of queries increased), we would need to shard the data across multiple volumes. This throws a lot of wrenches into our plans: we need to be able to automatically move data on increase or decrease of shard counts; we need to be able to route requests to the correct node with that data; we need to ensure that reconciliation between the data on disk and the metadata we store in Postgres is aware that there are multiple nodes where data _could_ reside.

Reviewer: ```suggestion
3. We locked ourselves into a node topology which is not easily scalable. We essentially have a singleton service over a persistent volume. In order to scale this service horizontally (which would become necessary once either the amount of data or the frequency of queries increased), we would need to shard the data across multiple volumes. This throws a lot of wrenches into our plans: we need to be able to automatically move data on increase or decrease of shard counts; we need to be able to route requests to the correct node with that data; we need to ensure that reconciliation between the data on disk and the metadata we store in Postgres is aware that there are multiple nodes where data _could_ reside.
```

4. There is no efficient way to implement certain data access patterns we may need in the future. For example, looking for data matching a certain identifier _name_ would require us to open every SQLite database on disk (and, no, _batching it_ is not a solution).

Instead of forging ahead and tackling the complexity we've made for ourselves, we decided to take a step back and re-evaluate whether or not our previous architecture choices were meeting our current needs. And it seems like they did not.

> Moving this data into Postgres seems like an obvious choice that will (eventually) be able to reduce or completely remove all of the burdens listed above.

Reviewer 2: "Seems like an obvious choice" and "(eventually)" sound wishy-washy and aren't confidence inspiring statements. Are you not confident this was the right decision?

Author: Will strengthen!

1. Reconciling pieces of data in the same datastore is trivial compared to reconciling pieces of data in two different datastores, especially if it should be done transactionally. Even though we're not in the same _physical_ database instance, having all of our data in _some_ Postgres-compatible store opens up the possibility of transactions spanning multiple physical Postgres instances via [postgres_fdw](https://www.postgresql.org/docs/12/postgres-fdw.html#id-1.11.7.42.12).
2. Moving data to an existing technology allows us to re-use the same advice and documentation for backup and disaster recovery strategies. Removing the number of knobs that must be turned by site-admins is always going to be a win across the board.
3. Moving from an embedded database to a server-client database removes the nasty singleton property on the service. Now that multiple servers can interact with the same data concurrently without risk of corruption, we can scale the services horizontally without requiring additional effort into implementing sticky sessions or complex request routing rules.

> 4. Moving data into a single relational store allows us to create additional indexed views into the data. We can now query over dimensions that were not possible before without opening tends of thousands of relatively small database files.

Reviewer 2: I think this section conflates reasons why we are making this change now, and potential benefits we will experience in the future. I think it is important to keep these things separate.

The conditional future tense used throughout this passage gives it won't be until some future point that we actually benefit from this change. As I reader I am left wondering "why make this change now?".

Author: I'll separate them (at least) and (maybe) remove the weaker one.

Reviewer 2: I think it is ok to talk about hypothetical future benefits as perks, as long as that is separate from and after the reasons why we are doing it now.

> ## Notes for site admins

Reviewer 2: Do we want this blog post to be the source of truth for information like this?

cc @christinaforney

Author: This information is already available in the upgrade docs and the changelog, I just wanted to call it out as an extra precaution.

Author: We already did have a question in a support channel about these values, so it's apparently pretty easy to glaze over some of the release notes.

Reviewer 2: I think the right way to call it out is to link to a source of truth. If those documents don't have sufficient information (customers have asked questions) then I think we should add context to the source of truth, not add more information here in this blog post.

Author: There are already links to https://docs.sourcegraph.com/admin/external_database and https://docs.sourcegraph.com/admin/updates in this text.

As for the question we got already:

```
does this use the pgsql instance we configured?  it looks like it is trying to stand up another db for this.... if that is the case, i need to find the settings to configure it to use the HA pgsql instance we have placed
```

I think the addition just took them by surprise, because their intuition about how to update their deployment was correct.

Here is the text in the upgrading section:

```
This release introduces a second database instance, codeintel-db. If you have configured Sourcegraph with an external database, then update the CODEINTEL_PG* environment variables to point to a new external database as described in the external database documentation. Again, these must not point to the same database or the Sourcegraph instance will refuse to start.
```

Reviewer 2: So my question is, why not delete all the content and just keep the links? If there is any content that you hesitate to delete, ask yourself if it belongs somewhere else (e.g. inside one of those links).

Author: The information under the data retention policy I think is useful, but would not fit nicely inside the changelog (where the parameter is introduced).

Maybe this isn't the best place for it, but I don't think the changelog is the best places for it either.

Reviewer 3: Let me know if you need more input on this post (I skimmed through it, and my concerns are the same that Nick mentions above - making sure the context is in the appropriate location).

Author: I'm assuming Nick and I will discuss this online in our 1-1 tomorrow, but I'll tap you if we discuss it earlier in case you're interested.

> If you are an active user of LSIF, there are a few changes that may subtly affect the behavior of the code intelligence backend. Our enterprise users operate at incredibly different scales and is difficult to ensure that a change has zero impact on all existing customers.

Reviewer: ```suggestion
If you are an active user of LSIF, there are a few changes that may subtly affect the behavior of the code intelligence backend. Our enterprise users operate at incredibly different scales so it is difficult to ensure that a change has zero impact on all existing customers.
```

The [3.20 -> 3.21 upgrade guides](https://docs.sourcegraph.com/admin/updates) call out instructions on how to do the same for the new code intelligence database. In the same way you configure your primary database instance with `PG*` environment variables, you configure the code intelligence database instance with `CODEINTEL_PG*` environment variables.

**Note** that if you configure both databases externally, they **must not** point to the exact same database. The frontend service will refuse to start with such a configuration.

#### Data retention policy

The [3.21 changelog](https://github.com/sourcegraph/sourcegraph/blob/3.21/CHANGELOG.md) details the introduction of a new environment variable, `PRECISE_CODE_INTEL_MAX_DATA_AGE`, on the precise-code-intel-bundle-manager service.

Previously, the bundle manager stored all LSIF data on disk as SQLite files. This allowed us to run custodial processes which removed old data, but only when disk pressure was high. Since we've migrated this data to a Postgres instance, there is a less clear-cut method to determine how much data we need to free in order to keep the desired amount of headroom.

> Disks do not grow on trees and we at Sourcegraph are well aware of that fact. In order to ensure we didn't accidentally slip an infinite disk space requirement to the Sourcegraph 3.21 release, we introduced an alternate data retention policy to replace the old full-disk trigger. Any data that is older than `PRECISE_CODE_INTEL_MAX_DATA_AGE` and is not used to serve code intelligence queries for the tip of a default branch will be automatically flushed from the code intelligence database periodically.

Reviewer: ```suggestion
Disks do not grow on trees and we at Sourcegraph are well aware of that fact. In order to ensure we didn't accidentally require an infinite disk space in the Sourcegraph 3.21 release, we introduced an alternate data retention policy to replace the old full-disk trigger. Any data that is older than `PRECISE_CODE_INTEL_MAX_DATA_AGE` and is not used to serve code intelligence queries for the tip of a default branch will be automatically flushed from the code intelligence database periodically.
```

> We believe that this will give us roughly the same behavior as the previous versions, and can be tuned to match expected behavior at your instance's scale. The default value for this environment variable is `720h` (30 days), which is likely adequate for the majority of instances. This value should be adjusted if the size of your index uploads are large and your commit frequency is high.

Reviewer: ```suggestion
We believe that this will give us roughly the same behavior as the previous versions, and can be tuned to match expected behavior at the scale of your instance. The default value for this environment variable is `720h` (30 days), which is likely adequate for the majority of instances. This value should be adjusted if the size of your index uploads are large and your commit frequency is high.
```

Using SQLite for this data has become an operational and feature burden for the code intelligence team.

- We need to maintain a periodic process to reconcile the data on disk with the metadata we store in Postgres.
- We have good instructions and tips for backup and recovery of our Postgres data store, but have never officially recommended backing up other persistent volumes.
- We locked ourselves into a node topology which is not easily scalable. We essentially have a singleton service over a persistent volume. In order to scale this service horizontally (which would become necessary once either the amount of data or the frequency of queries increased), we would need to shard the data across multiple volumes. This throws a lot of wrenches into our plans: we need to be able to automatically move data on increase or decrease of shard counts; we need to be able to route requests to the correct node with that data; we need to ensure that reconciliation between the data on disk and the metadata we store in Postgres is aware that there are multiple nodes where data _could_ reside.

> - There is no efficient way to implement certain data access patterns we may need in the future. For example, looking for data matching a certain identifier _name_ would require us to open every SQLite database on disk (and, no, _batching it_ is not a solution).

Reviewer 4: wdym by batching in this context?

Author: Clarified.

Instead of forging ahead and tackling the complexity we've made for ourselves, we decided to take a step back and re-evaluate whether or not our previous architecture choices were meeting our current needs. And it seems like they did not.

Moving this data into Postgres was an obvious choice that greatly reduced the burdens listed above.

> - Moving data to an existing technology allows us to re-use the same advice and documentation for backup and disaster recovery strategies. Removing the number of knobs that must be turned by site-admins is always going to be a win across the board.

Reviewer 4: Change `existing` to something like "already used elsewhere in our tech stack"?

Reviewer 4: also, either 'removing knobs' or 'lowering the number of knobs'

Reviewer 4: also, did this change reduce the number of knobs, or just the number of types of knobs?

Author: #1: Updated.
#2: Good catch.
#3: Net number of knobs is lower. We'll no longer have to worry about resource allocation in the bundle manager query path, and we'll be able to remove most of the janitor settings (once we get S3/GCS/MinIO to replace the rest of the bundle manager, there will be like 1-2 settings at all for this).
